### PR TITLE
Jenkins 48018

### PR DESCRIPTION
**Pull request by ndeloof**
See [JENKINS-48018](https://issues.jenkins-ci.org/browse/JENKINS-48018).

### Proposed changelog entries

* add support for data-binding on Descriptors

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (developer)
- [ ] Appropriate explanation to why this change has no tests:
       need some recommandation on how to write simple test involving StaplerRequest

### Desired reviewers

@jglick 
@jenkinsci/code-reviewers

Note: primary goal is to provide discovery of Descriptor attributes relying on `@DataBound*`. see https://github.com/jenkinsci/configuration-as-code-plugin

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Some autotests are required at least

**Comment by ndeloof** [comment]
@oleg-nenashev yes, as noted I need some recommandation. I tried to define a fake Descriptor but then need a fake StaplerRequest ... and this results in significant test setup

File: core/src/main/java/hudson/model/Descriptor.java [hunk 1]
```diff
@@ -802,8 +806,61 @@ public boolean configure( StaplerRequest req ) throws FormException {
...
+     */
+    public void reset() {
+        for (PropertyDescriptor pd : PropertyUtils.getPropertyDescriptors(this)) {
+            final Method writeMethod = pd.getWriteMethod();
+            if (writeMethod != null && writeMethod.getAnnotation(DataBoundSetter.class) != null) {
+                try {
+                    final Class cl = writeMethod.getParameterTypes()[0];
+                    Object value;
+                    try {
+                        final Field field = getClass().getField(pd.getName() + "_default");
```
Comment (daniel-beck) [inline core/src/main/java/hudson/model/Descriptor.java]: Can't we create an instance and check what the getters return, similar to what the Pipeline Snippet Generator does?

Comment (ndeloof) [reply core/src/main/java/hudson/model/Descriptor.java]: yes indeed, but relying on such convention allows to get the jelly tags `default` set as well.

**Comment by daniel-beck** [comment]
What is the status of this? Are you seeking to have this merged into master? If so, could you provide the downstream PR and use case?

(Since https://github.com/jenkinsci/system-config-dsl-plugin didn't get beyond the prototype stage, I'd rather not have a repeat of https://github.com/jenkinsci/jenkins/pull/2293)

**Comment by ndeloof** [comment]
@daniel-beck this is not for system-config-dsl-plugin  but https://github.com/jenkinsci/configuration-as-code-plugin (as noted on description)

use-case is, I guess, well described on https://issues.jenkins-ci.org/browse/JENKINS-48018 with link to mailing list discussion. Maybe https://github.com/jenkinsci/configuration-as-code-plugin/issues/2 could help for understanding.

and downstream PR [Jenkins 48018 #38](https://github.com/jenkinsci/mailer-plugin/pull/38) is referenced as well, do you expect a special format for this purpose ?

**Comment by daniel-beck** [comment]
@ndeloof Thanks for the explicit references.

**Comment by oleg-nenashev** [comment]
@ndeloof Is it still actual? IIUC https://github.com/stapler/stapler/pull/140 does similar thing in a better way

**Comment by ndeloof** [comment]
I indeed think this is obsolete with a general databinding enhancement on his way

**Pull request closed without merging**